providers/jdbc/shedlock-provider-jdbc-template/src/main/java/net/javacrumbs/shedlock/provider/jdbctemplate/JdbcTemplateStorageAccessor.java

```diff
@@ -15,47 +15,74 @@
  */
 package net.javacrumbs.shedlock.provider.jdbctemplate;
 
-import net.javacrumbs.shedlock.provider.jdbc.internal.AbstractJdbcStorageAccessor;
-import org.springframework.dao.DataAccessException;
+import net.javacrumbs.shedlock.core.LockConfiguration;
+import net.javacrumbs.shedlock.support.AbstractStorageAccessor;
 import org.springframework.dao.DataIntegrityViolationException;
-import org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator;
-import org.springframework.jdbc.support.SQLExceptionTranslator;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
 
-import javax.sql.DataSource;
-import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.time.Instant;
+
+import static java.util.Objects.requireNonNull;
 
 /**
- * Spring JdbcTemplate based implementation which at the end does not use much from JdbcTemplate except
- * exception translator. The reason is that we rely on atomic JDBC operations and thus do not want to
- * participate in transactions. It's not easy to not use transactions with JdbcTemplate and since we were using
- * only exception translation, it makes sense to skip JdbcTemplate completely.
+ * Spring JdbcTemplate based implementation usable in JTA environment
  */
-class JdbcTemplateStorageAccessor extends AbstractJdbcStorageAccessor {
-    private final SQLExceptionTranslator exceptionTranslator;
-
-    JdbcTemplateStorageAccessor(DataSource dataSource, String tableName) {
-        super(dataSource, tableName);
-        this.exceptionTranslator = new SQLErrorCodeSQLExceptionTranslator(dataSource);
-    }
-
-    @Override
-    protected void handleInsertionException(String sql, SQLException e) {
-        DataAccessException translatedException = exceptionTranslator.translate("InsertLock", sql, e);
-        if (translatedException instanceof DataIntegrityViolationException) {
-            // lock record already exists
-            // DuplicateKeyException is not enough for Vertica
-        } else {
-            throw translatedException;
-        }
-    }
-
-    @Override
-    protected void handleUpdateException(String sql, SQLException e) {
-        throw exceptionTranslator.translate("UpdateLock", sql, e);
-    }
-
-    @Override
-    protected void handleUnlockException(String sql, SQLException e) {
-        throw exceptionTranslator.translate("Unlock", sql, e);
-    }
+class JdbcTemplateStorageAccessor extends AbstractStorageAccessor {
+	private final String tableName;
+	private final JdbcTemplate jdbcTemplate;
+
+	JdbcTemplateStorageAccessor(JdbcTemplate jdbcTemplate, String tableName) {
+		this.jdbcTemplate = requireNonNull(jdbcTemplate, "jdbcTemplate can not be null");
+		this.tableName = requireNonNull(tableName, "tableName can not be null");
+	}
+
+	@Override
+	@Transactional(propagation = Propagation.REQUIRES_NEW)
```

Comment:
Are you sure that the annotation will get processed? (I am afraid that it will not be scanned as it is in a library package) TransactionalTemplate might be safer choice

Reply:
I have changed the code to use TransactionTemplate (added constructor parameter for injecting a PlatformTransactionManager, if not set, a DataSourceTransactionManager is created and used).